File: android/app/src/main/kotlin/com/usesmileid/sampleapps/android/navigation/TokenDestinations.kt
package com.usesmileid.sampleapps.android.navigation

import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.remember
import androidx.compose.runtime.saveable.rememberSaveable
import androidx.compose.runtime.setValue
import androidx.compose.ui.platform.LocalClipboardManager
import com.ramcosta.composedestinations.annotation.Destination
import com.ramcosta.composedestinations.annotation.RootGraph
import com.ramcosta.composedestinations.annotation.parameters.DeepLink
import com.ramcosta.composedestinations.generated.destinations.ScanTokenScreenDestination
import com.ramcosta.composedestinations.generated.destinations.SdkFlowScreenDestination
import com.ramcosta.composedestinations.navigation.DestinationsNavigator
import com.usesmileid.sampleapps.android.LocalUseSmileIDSampleAppState
import com.usesmileid.sampleapps.android.flow.UseSmileIDSampleFlowTokens
import com.usesmileid.sampleapps.android.scan.UseSmileIDSampleQrScanner
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleRunIntent
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleTokenDecoder
import com.usesmileid.sampleapps.ui.state.UseSmileIDSampleTokenSession
import kotlinx.coroutines.launch
import com.usesmileid.sampleapps.ui.screens.ScanTokenScreen as ScanTokenContent

/** The token-scanning route. Function names are load-bearing: KSP names each generated `…Destination` after the function. */

@Destination<RootGraph>(deepLinks = [DeepLink(uriPattern = UseSmileIDSampleDeepLinks.SCAN_TOKEN)])
@Composable
fun ScanTokenScreen(navigator: DestinationsNavigator) {
    val app = LocalUseSmileIDSampleAppState.current
    val clipboard = LocalClipboardManager.current
    var torchOn by rememberSaveable { mutableStateOf(false) }
    // Claimed on arrival so it belongs to this visit: leaving by any route drops it with the screen,
    // and a later unrelated scan cannot resurrect a run the partner has forgotten about.
    val claimed = rememberSaveable { app.interruptedRun.claim()?.saved() ?: emptyList() }
    val resuming = remember(claimed) { UseSmileIDSampleRunIntent.of(claimed) }
    // One path for both entry routes — typed, pasted or simulated, a session is linked the same way.
    val link: (UseSmileIDSampleTokenSession) -> Unit = { session ->
        // On the app-level scope, so leaving this screen cannot cancel the write half-done.
        app.storeScope.launch { app.store.linkTokenSession(session) }
        // A resumed run leaves on the effect below instead, once the linked session is actually live.
        if (resuming == null) navigator.navigateUp()
    }
    // Waits for the store's write to reach the session state rather than navigating on the link:
    // re-entering the flow against the expired session would fail the gate and bounce straight back.
    // A relinked span that is itself expired never satisfies this, which is the loop's other half.
    // Keyed on the session rather than on `sessionActive`, whose clock read would recompose this
    // screen — and rebind its camera — once a second for as long as a session is live.
    val current = app.session
    LaunchedEffect(resuming, current) {
        if (resuming != null && current?.hasExpired(System.currentTimeMillis()) == false) {
            navigator.navigate(SdkFlowScreenDestination(productId = resuming.productId, route = resuming.route)) {
                popUpTo(ScanTokenScreenDestination) { inclusive = true }
            }
        }
    }
    ScanTokenContent(
        // Why this screen opened, per R10: the redirect's message belongs to the screen it arrives at.
        reason = resuming?.let { SESSION_ENDED_REASON },
        onBack = { navigator.navigateUp() },
        onLink = link,
        onPaste = { clipboard.getText()?.text },
        onSimulate = { span, bindings ->
            val minted = UseSmileIDSampleFlowTokens.session(
                span = span,
                bindings = bindings,
                nowMillis = System.currentTimeMillis(),
            )
            // The minter and the decoder have to agree, and a fixture that no longer decodes is a defect
            // rather than something to paper over with a fabricated session.
            UseSmileIDSampleTokenDecoder.session(minted)?.let(link)
        },
        torchOn = torchOn,
        onTorchToggle = { torchOn = !torchOn },
        // The camera lives in the shell: `sample-ui` runs under eight identities, and only this one
        // owns a scanner. It unbinds on leaving composition, so the SDK gets the camera back (§7.1).
        viewfinder = { modifier, enabled, onCandidate ->
            UseSmileIDSampleQrScanner(
                onCode = onCandidate,
                torchOn = torchOn,
                enabled = enabled,
                modifier = modifier,
            )
        },
    )
}

/** Product string for the expiry redirect, kept beside the redirect rather than in the shared screen. */
private const val SESSION_ENDED_REASON = "Token session ended. Scan to continue where you left off."

File: ios/App/Sources/Navigation/TokenDestinations.swift
import AVFoundation
import SampleUI
import SwiftUI

/// The token-scanning route: the shared screen, given the shell's camera, clipboard and minter.
struct UseSmileIDSampleScanTokenHost: View {
  @EnvironmentObject private var router: UseSmileIDSampleRouter
  @EnvironmentObject private var app: UseSmileIDSampleAppState
  @State private var torchOn = false

  /// Claimed on arrival, so leaving by any route drops it and a later scan resurrects nothing.
  @State private var resuming: UseSmileIDSampleRunIntent?
  /// So the resume waits for a *different* session.
  @State private var arrivedWith: String?
  @State private var resumeHandled = false

  var body: some View {
    ScanTokenScreen(
      entry: $app.scanEntry,
      // R10: the redirect's message belongs to the screen it arrives at.
      reason: resuming == nil ? nil : .sessionEnded,
      torchOn: torchOn,
      onBack: { router.pop() },
      onLink: link,
      onSimulate: { span, bindings, environment in
        let minted = UseSmileIDSampleFlowTokens.session(span: span, bindings: bindings, environment: environment, now: Date())
        // The minter and decoder have to agree: a fixture that no longer decodes is a defect, not a session to fabricate.
        if let session = UseSmileIDSampleTokenDecoder.session(minted) {
          link(session)
        }
      },
      // Read behind the tap and never earlier: iOS 16+ prompts on a read the person did not initiate.
      onPaste: { UIPasteboard.general.string },
      onTorchToggle: { torchOn.toggle() },
      viewfinder: viewfinder
    )
    .onAppear(perform: claim)
    // Not `sessionActive`, whose clock read would fire once a second.
    .onChange(of: app.session?.id) { _ in resume() }
  }

  /// The camera lives in the shell, `SampleUI` running under eight identities; absent, the screen keeps the glyph.
  private var viewfinder: UseSmileIDSampleViewfinder? {
    guard AVCaptureDevice.default(.builtInWideAngleCamera, for: .video, position: .back) != nil else { return nil }
    let torchOn = torchOn
    return { enabled, onCandidate in
      AnyView(UseSmileIDSampleQrScanner(onCode: onCandidate, torchOn: torchOn, enabled: enabled))
    }
  }

  private func claim() {
    guard resuming == nil else { return }
    arrivedWith = app.session?.id
    resuming = app.interruptedRun
    app.interruptedRun = nil
  }

  /// Typed, pasted or simulated, a session is linked the same way.
  private func link(_ session: UseSmileIDSampleTokenSession) {
    app.linkSession(session)
    // A resumed run leaves on the change below, once the write has reached the app state.
    if resuming == nil {
      router.pop()
    }
  }

  /// Only on a *different* session: re-entering against the one that sent us here would bounce back.
  private func resume() {
    guard let resuming, !resumeHandled, let linked = app.session, linked.id != arrivedWith else { return }
    resumeHandled = true
    guard !linked.hasExpired(at: Date()) else {
      // An expired relink cannot start the run, and freezing on "linked" says nothing.
      router.pop()
      return
    }
    // Opened, not pushed: the path is assigned, so two quick links cannot stack two runs.
    router.open(.sdkFlow(productId: resuming.productId, presentation: resuming.route))
  }
}
